### PR TITLE
feat(http): support CSRF provention

### PR DESCRIPTION
**Pull request by GeorgeMac**
Fixes FLI-116

This adds support for a configurable CSRF key to be added and protection to be enforced.

When configured (non-empty `authentication.crsf.key`) a CSRF credential will be configured and stored in a cookie when accessing any of the following API prefixes:

- /api/v1
- /auth/v1
- /meta

Attempts to use non-safe (all bar GET or HEAD) HTTP methods will require an acquired CSRF token be presented.
The token is verifiable via the CSRF credentials stored in the cookie (signature verification is used).

The token can be acquired via a response header obtained on any route beneath `/meta`.
The UI performs a `/meta/info` request in each load.
This is where the current UI is obtaining it and storing it in its Vuex store.

The UI has been updated to obtain the token and present it on all subsequent API requests.
The new UI will need a similar adjustment to be made to support CSRF.

This method of prevention ensures that a malicious request cannot be crafted from another domain / origin and directed over to Flipt to perform a malicious operation (e.g. delete or update a flag or authentication).

There is one caveat, and that is the enforcement is only performed when the request contains an `Origin` header.
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin

This is taken from the MDN link above:

---

Broadly speaking, user agents add the [Origin](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin) request header to:

- [cross origin](https://developer.mozilla.org/en-US/docs/Glossary/CORS) requests.
- [same-origin](https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy) requests except for [GET](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/GET) or [HEAD](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/HEAD) requests (i.e. they are added to same-origin [POST](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/POST), [OPTIONS](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/OPTIONS), [PUT](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PUT), [PATCH](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PATCH), and [DELETE](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/DELETE) requests).

---

This ensures we enforce the policy for browser user-agents. However, not for other clients (e.g. our SDKs).
Thereby avoiding having to teach clients to perform the same CSRF prevention, which should be an unnecessary security step for our SDKs. As, the attack is performed by manipulating a user's browser and cookie-based authentication.

### CODE REVIEW

File: internal/config/authentication.go [hunk 1]
```diff
@@ -123,6 +123,14 @@ type AuthenticationSession struct {
...
 	StateLifetime time.Duration `json:"stateLifetime,omitempty" mapstructure:"state_lifetime"`
+	// CSRF configures CSRF provention mechanisms.
+	CSRF AuthenticationSessionCSRF `json:"csrf,omitempty" mapstructure:"csrf"`
+}
+
+// AuthenticationSessionCSRF configures cross-site request forgery prevention.
+type AuthenticationSessionCSRF struct {
+	// Key is the private key string used to authenticate csrf tokens.
+	Key string `json:"-" mapstructure:"key"`
 }
```
Comment (markphelps) [inline internal/config/authentication.go:134]: wonder if we should add an IT that verifies that this key is never exposed (if configured) when curling `/meta/config` ?

Comment (GeorgeMac) [reply internal/config/authentication.go:134]: That is a good idea. To avoid this regressing.

Comment (GeorgeMac) [reply internal/config/authentication.go:134]: I updated the test to both look for the key that was configured in the meta config response and ensure that the CSRF `Set-Cookie` header was supplied appropriately.

**Review by markphelps — APPROVED** [review]
⚡️

**Pull request merged**